main/src/ca/uwaterloo/flix/tools/Tester.scala

```diff
@@ -1,125 +1,208 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
 package ca.uwaterloo.flix.tools
 
+import ca.uwaterloo.flix.api.Flix
 import ca.uwaterloo.flix.language.ast.Symbol
 import ca.uwaterloo.flix.runtime.CompilationResult
-import ca.uwaterloo.flix.util.Formatter
+import ca.uwaterloo.flix.util.{Duration, InternalCompilerException, TimeOps}
+import org.jline.terminal.{Terminal, TerminalBuilder}
+
+import java.util.concurrent.ConcurrentLinkedQueue
+import java.util.logging.{Level, Logger}
 
 /**
- * Evaluates all tests in a model.
- */
+  * Evaluates all tests in a Flix program.
+  */
 object Tester {
 
   /**
-   * Represents the outcome of a single test.
-   */
-  sealed trait TestResult {
-    /**
-     * The symbol associated with the test.
-     */
-    def sym: Symbol.DefnSym
-  }
-
-  object TestResult {
-
-    /**
-     * Represents a successful test case.
-     */
-    case class Success(sym: Symbol.DefnSym, msg: String) extends TestResult
-
-    /**
-     * Represents a failed test case.
-     */
-    case class Failure(sym: Symbol.DefnSym, msg: String) extends TestResult
-
+    * Runs all tests.
+    */
+  def run(compilationResult: CompilationResult)(implicit flix: Flix): Unit = {
+    //
+    // Find all test cases (both active and ignored).
+    //
+    val tests = getTestCases(compilationResult)
+
+    // Start the TestRunner and TestReporter.
+    val queue = new ConcurrentLinkedQueue[TestEvent]()
+    val reporter = new TestReporter(queue, tests)
+    val runner = new TestRunner(queue, tests)
+    reporter.start()
+    runner.start()
+
+    // Wait for everything to complete.
+    reporter.join()
+    runner.join()
   }
 
   /**
-   * Represents the outcome of a run of a suite of tests.
-   */
-  sealed trait OverallTestResult
-
-  object OverallTestResult {
+    * A class that reports the results of test events as they come in.
+    */
+  private class TestReporter(queue: ConcurrentLinkedQueue[TestEvent], tests: Vector[TestCase])(implicit flix: Flix) extends Thread {
+    override def run(): Unit = {
+      // Silence JLine warnings about terminal type.
+      Logger.getLogger("org.jline").setLevel(Level.OFF)
+
+      // Initialize the terminal.
+      implicit val terminal: Terminal = TerminalBuilder
+        .builder()
+        .system(true)
+        .build()
+      val writer = terminal.writer()
+
+      // Print headline.
+      writer.println(s"Running ${tests.length} tests...")
+      writer.println()
+      writer.flush()
+
+      // Main event loop.
+      var passed = 0
+      var failed = 0
+
+      var finished = false
+      while (!finished) {
+        queue.poll() match {
+          case TestEvent.Success(sym, elapsed) =>
```

Comment:
A minor suggestion for your consideration:

It would be nice to show the active test, _then_ show whether or not it succeeded.
For example:
```
testDoesTheCoolStuff // no newline
// wait half a second
OK
```
so we end up with
```
test01 OK
test02 OK
test03 OK
test04
```

And we see we're waiting on test04. Maybe that one causes an infinite loop?

Reply:
Maybe that doesn't work well with concurrency though

Reply:
Agreed. Upcoming.